### PR TITLE
Model fixes

### PR DESCRIPTION
**Pull request by acburdine**
This is a regression introduced in the container-injections branch. Will try and figure out some way to test this.

### CODE REVIEW

**Comment by acburdine** [comment]
This actually needs further thought, as after this fix the initializer errors out on not being able to find an adapter for the specific model

**Comment by davewasmer** [comment]
What is the actual original bug you are trying to fix here?

**Comment by acburdine** [comment]
There wasn't necessarily an original bug. There was the change I made a while back fixing the issue where containers would break the `hasOwnProperty` functionality. The specific case where that came up was in the idea of abstract models (e.g. an "application" model) that doesn't actually need to be used as an actual model. The `hasOwnProperty` check was removed during the container refactor, so I re-added it.

Also, there was some repeated behavior where the define-orm-models initializer re-implemented the adapter lookup functionality that also exists in the model, so I made it use the model's getAdapter method to lookup rather than repeat the same code.

The last thing is the addition of the `create` method in the database service. It was missing 😛

**Comment by acburdine** [comment]
So essentially the "regression" I was referring to was fixed in the first commit, then I added the second commit because it also related to models 🤷‍♂️

**Comment by acburdine** [comment]
This should be good to go as well.

File: config/initializers/define-orm-models.ts [hunk 1]
```diff
@@ -14,14 +14,14 @@ export default {
...
     let modelsGroupedByAdapter = new Map();
-    forEach(models, (ModelClass, modelName) => {
-      if (ModelClass.abstract) {
+    forEach(models, (ModelClass: typeof Model) => {
+      if (ModelClass.hasOwnProperty('abstract') && ModelClass.abstract) {
         return;
       }
-      let Adapter = container.lookup(`orm-adapter:${ modelName }`, { loose: true }) || container.lookup('orm-adapter:application');
+      // Let the model class resolve its own adapter
+      let Adapter = ModelClass.getAdapter(application.container);
```
Comment (davewasmer) [inline config/initializers/define-orm-models.ts]: Not sure about this - especially if we move the model query methods to the database service (as discussed).

I see some value in letting a Model class determine it's own adapter, and the rest of the system respecting that. But on the other hand, is that level of customization really useful? It kinda feels like it's defeating the whole point of the container's conventions.

I'm on the fence, curious to hear what others think.

File: lib/data/model.ts [hunk 2]
```diff
@@ -124,13 +124,54 @@ export default class Model extends DenaliObject {
...
     }
     assert(adapter, `No orm-adapter found for "${ this.type }", and no fallback "application" orm-adapter found either. Available adapters: ${ container.availableForType('orm-adapter') }`);
     return adapter;
   }
 
   /**
+   * Call the supplied callback function for each attribute on this model, passing in the attribute
+   * name and attribute instance.
+   */
+  static mapAttributes<T>(container: Container, fn: (descriptor: Attribute, attributeName: string) => T): T[] {
```
Comment (davewasmer) [inline lib/data/model.ts]: Why make these static?

File: lib/data/model.ts [hunk 3]
```diff
@@ -283,8 +324,8 @@ export default class Model extends DenaliObject {
    * attributes
    */
   inspect(): string {
-    let attributesSummary: string[] = this.mapAttributes((value, attributeName) => {
-      return `${ attributeName }=${ JSON.stringify(value) }`;
+    let attributesSummary: string[] = (<typeof Model>this.constructor).mapAttributes(this.container, (value, attributeName) => {
+      return `${ attributeName }=${ JSON.stringify(this[attributeName]) }`;
```
Comment (davewasmer) [inline lib/data/model.ts]: Why not use `value` here?

File: lib/data/model.ts (hunk 2 above)
Comment (acburdine) [inline lib/data/model.ts]: So the reason for making them static is twofold:

1) ORM adapters need to be able to loop through attributes and relationships on the model class in order to get set up correctly. This is the only way of doing it. The orm-adapter could create an instance of each model, but even then it wouldn't have access to the actual attribute/relationship descriptor unless it re-implemented the static behavior
2) An instance can call the static method, then get the instance value using the key of the attribute/relationship, but an orm can't infer the attribute descriptor from the instance value.

File: lib/data/model.ts (hunk 3 above)
Comment (acburdine) [inline lib/data/model.ts]: see above. When they're made static, the value itself is actually the attribute descriptor, so we need to use `this[attributeName]` to get the actual instance value

File: config/initializers/define-orm-models.ts (hunk 1 above)
Comment (acburdine) [inline config/initializers/define-orm-models.ts]: Personally I didn't do it from a customization perspective, more-so from a "dry up the code" approach, but will drop this change for now and can add it back in later.

**Comment by acburdine** [comment]
Removed all of the unnecessary changes with the model refactor; this should be 👍 now.

**Pull request merged**